File: 347-top-k-frequent-elements/top-k-frequent-elements.kt
class Solution {
    fun topKFrequent(nums: IntArray, k: Int): IntArray {
        val freqMap = mutableMapOf<Int, Int>()
        nums.forEach { freqMap[it] = freqMap.getOrDefault(it, 0) + 1 }

        // Initialize bucket (index = frequency)
        val buckets = Array<MutableList<Int>>(nums.size + 1) { mutableListOf() }
        for ((num, freq) in freqMap) {
            buckets[freq].add(num)
        }

        val result = mutableListOf<Int>()
        for (i in buckets.size - 1 downTo 0) {
            for (num in buckets[i]) {
                result.add(num)
                if (result.size == k) return result.toIntArray()
            }
        }

        return intArrayOf()
    }
}

File: 347-top-k-frequent-elements/top-k-frequent-elements.py
class Solution(object):
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        count = Counter(nums)
        res = [[] for _ in range(len(nums)+1)]
        out = []
        for value, freq in count.items():
            res[freq].append(value)
        for i in res[::-1]:
            for j in i:
                if k!=0:
                    out.append(j)
                    k -=1
                else:
                    return out
        return out
        